Plugins/Animation/src/AnimationComponent.cpp

```diff
@@ -1,397 +1,460 @@
-                                             m_boneDrawables.end(),
-                                             [roIdx]( const auto& bone) { return bone->getRenderObjectIndex() == roIdx; });
-
-        const uint boneIdx =  (*bonePos)->getBoneIndex();
-        const Ra::Core::Transform& TBoneModel = m_skel.getTransform(boneIdx, Ra::Core::Animation::Handle::SpaceType::MODEL);
-        const Ra::Core::Transform& TBoneLocal = m_skel.getTransform(boneIdx, Ra::Core::Animation::Handle::SpaceType::LOCAL);
-        auto diff = TBoneModel.inverse() *  transform;
-        m_skel.setTransform(boneIdx,TBoneLocal * diff,  Ra::Core::Animation::Handle::SpaceType::LOCAL);
+        return Scalar( 0 );
     }
+    return m_animations[m_animationID].getDuration();
+}
 
-    uint AnimationComponent::getBoneIdx(Ra::Core::Index index) const
+uint AnimationComponent::getMaxFrame() const {
+    if ( m_animations.empty() )
     {
-        auto found = std::find_if(m_boneDrawables.begin(), m_boneDrawables.end(),
-                                  [index](const auto& draw) { return draw->getRenderObjectIndex() == index; });
-        return found == m_boneDrawables.end() ?
-               uint(-1) :
-               (*found)->getBoneIndex();
+        return 0;
     }
+    return uint( std::round( getDuration() / m_dt[m_animationID] ) );
+}
 
-    const Ra::Core::Animation::Animation* AnimationComponent::getAnimation() const
+void AnimationComponent::cacheFrame( const std::string& dir, int frame ) const {
+    std::ofstream file( dir + "/" + m_contentName + "_frame" + std::to_string( frame ) + ".anim",
+                        std::ios::trunc | std::ios::out | std::ios::binary );
+    if ( !file.is_open() )
     {
-        if (m_animations.empty())
-        {
-            return nullptr;
-        }
-        return &m_animations[m_animationID];
+        return;
     }
+    file.write( (const char*)&m_animationID, sizeof( uint ) );
+    file.write( (const char*)&m_animationTimeStep, sizeof( bool ) );
+    file.write( (const char*)&m_animationTime, sizeof( Scalar ) );
+    file.write( (const char*)&m_speed, sizeof( Scalar ) );
+    file.write( (const char*)&m_slowMo, sizeof( bool ) );
+    const auto& pose = m_skel.getPose( Handle::SpaceType::LOCAL );
+    file.write( (const char*)pose.data(), sizeof( Ra::Core::Transform ) * pose.size() );
+    LOG( logINFO ) << "Saving anim data at time: " << m_animationTime;
+}
 
-    const Scalar* AnimationComponent::getTimeOutput() const
+bool AnimationComponent::restoreFrame( const std::string& dir, int frame ) {
+    std::ifstream file( dir + "/" + m_contentName + "_frame" + std::to_string( frame ) + ".anim",
+                        std::ios::in | std::ios::binary );
+    if ( !file.is_open() )
     {
-        return &m_animationTime;
+        return false;
     }
-
-    Scalar AnimationComponent::getTime() const
+    file.read( (char*)&m_animationID, sizeof( uint ) );
```

Comment:
Well, the loop is in the caller `AnimationSystem::restoreFrame`. And The buffer size is ensured here. Perhaps by writing `file.read(reinterpret_cast<char*>(&m_animationID), sizeof n))`, as in the documentation https://en.cppreference.com/w/cpp/io/basic_istream/read it would helps.
@hoshiryu please try in your next PR about animation.
Same for others below.